### ISSUE TITLE
Root/sudo permission

### ISSUE THREAD

**Issue by ricardorlg** [gitpod-io/gitpod#39]
Just a question, I'm trying to execute some selenium tests, I was trying to add chrome but looks like the user is not sudo in the container.

**Comment by svenefftinge** [comment]
Yes, at runtime the workspace container runs without sudo or root for security reasons.
But you can install any tools in custom Docker image and set it in your `.gitpod` file.
Mor information here: http://docs.gitpod.io/40_Configuration.html#docker-image

**Comment by svenefftinge** [comment]
> It might be worth revisiting these security reasons, as in older Docker versions, root was able to escape containers on to the host, but that was fixed a few years ago and is no longer true in recent Docker versions.

Do you have a reference for this claim? We are waiting for a recent Docker change to land in Kubernetes.

**Comment by jankeromnes** [comment]
> Do you have a reference for this claim? We are waiting for a recent Docker change to land in Kubernetes.

Sorry, I don't have a solid reference for this, and looking around I see that most of the security best practices for Docker and [Kubernetes](https://kubernetes.io/docs/tasks/administer-cluster/securing-a-cluster/#controlling-what-privileges-containers-run-with) still mention restricting root access as a good idea.

From my understanding this is a "defense in depth" thing. Today I don't think that there are obvious ways to use `sudo` to escape a container (with a recent Docker on a recent kernel), but historically escapes may have been easier to achieve with `sudo` than without (e.g. using kernel exploits; forcing namespace changes; chrooting to a host directory; etc).

Nowadays, "root" privileges are segmented into capabilities (which can be granted or not using a seccomp profile, in order to enable valid use cases while preventing dangerous exploits), and modern Linux namespaces effectively segregate users and resources inside a container from host-level stuff, but I'm by no means a security expert and maybe it's still too risky to allow `sudo` inside containers.

**Comment by svenefftinge** [comment]
We definitely want to and will allow sudo at runtime. But as mentioned earlier we are awaiting some announced changes in the technologies we use. 
At the same we will make it easier configure a custom workspace with the tools you need, so you don't need to add additional things at runtime.

**Comment by svenefftinge** [comment]
Until https://github.com/kubernetes/kubernetes/pull/64005 has landed we could consider whitelisting certain trusted users to allow sudo/root.

**Comment by stuartpb** [comment]
Worth noting that there are multiple upstream approaches to this: https://github.com/kubernetes/enhancements/issues/127 tracks all potential PRs for it

**Comment by geropl** [comment]
Thanks @stuartpb. We're watching that thread closely for 1,5 years now, but:
 1. it's from 201**6**
 2. it always gets delayed.
It seems like the general trend is to push the problem to then actual container runtimes.

**Comment by kevinhq** [comment]
sudo: effective uid is not 0, is /usr/bin/sudo on a file system with the 'nosuid' option set or an NFS file system without root privileges?

Get this error on installling certain ruby version with rvm. Guess this IDE is not for me yet.

**Comment by kevinhq** [comment]
Nvm. Fixed with customization on Dockerfile.

**Comment by cmarqu** [comment]
@kevinhq What was the problem in the end, i.e. what was the customization you did?

**Comment by Kreyren** [comment]
Relevant: https://github.com/docker/compose/issues/7244#issuecomment-625834398

Development of rootless docker

**Comment by kevinhq** [comment]
sudo isn't needed since we can install everything through dockerfile

**Comment by Kreyren** [comment]
@kevinhq No we can't

**Comment by janat08** [comment]
Is the rootless version coming at some point? It sounds somewhat useless without apt.

**Comment by jankeromnes** [comment]
> Is the rootless version coming at some point?

Yes, it’s on the roadmap, although I can’t tell you when it will be released. Probably ‘sudo’ first at some point (allowing ‘sudo apt-get’ in workspace terminal, but still restricted by Docker’s default security model), and later rootless Docker cluster (allowing truly privileged containers with more capabilities).

> It sounds somewhat useless without apt.

You can already use ‘sudo apt-get’ in Gitpod like this: https://github.com/nushell/nushell/blob/master/.gitpod.Dockerfile

You can also use ‘brew install’ directly in a Gitpod terminal (no ‘sudo’ needed).

**Comment by Kreyren** [comment]
FWIW i am working on Proof-of-concept for gitpod to switch on VM-based containers which so far looks more economical and doesn't have this restriction + Allows VM support (also makes us to use other kernels)

Still needs some testing and security research..

> You can also use ‘brew install’ directly in a Gitpod terminal (no ‘sudo’ needed).

You can also use rootless apt which creates a new environment in your /home directory and adds the executables in PATH, i have janky implementation for this which i plan to implement if VM-based containers fails.

**Comment by dclong** [comment]
Will a recorded video be available later?

**Comment by jankeromnes** [comment]
@dclong Yes, with the livestream concluded, the link above is now a regular YouTube video.

**Comment by jankeromnes** [comment]
Update: You can now become `root` and use `sudo` in Gitpod! 🎉 (Beta feature)

Here is how to get `sudo` / `root` in your Gitpod workspace:

1. In https://gitpod.io/settings/ enable "Feature Preview" (but first read the description carefully)

2. Then create a new Gitpod workspace for any repository or branch

3. In that workspace, you should be able to use `sudo`, e.g. to run something like `sudo apt-get install mysql-server`

If you notice anything strange, please report it: You can add a comment here; or file a [new issue](https://github.com/gitpod-io/gitpod/issues/new/choose); or post feedback in https://community.gitpod.io -- And if in doubt, you can always disable "Feature Preview" again any time.

You can learn more about how this works in the [video](https://www.youtube.com/watch?v=l4I2TVAnBuw) above. We hope you'll enjoy this new feature as much as we do!

**Comment by pattacini** [comment]
That sounds awesome 🚀

**Comment by hendramn** [comment]
Nice!

**Comment by Kreijstal** [comment]
wow, this can actually be used now

**Comment by kevinhq** [comment]
here's hoping it will be the public feature soon (not beta anymore).

**Comment by kevinhq** [comment]
not working with default docker image from debian 10.  it's not permitted to install sudo?

```
building base image

initializing base image context
UGit operations complete                       location=/workspace/init
running base image build
Step 1/3 : FROM debian:10
10: Pulling from library/debian
Digest: sha256:9d4ab94af82b2567c272c7f47fa1204cd9b40914704213f1c257c44042f82aac
Status: Image is up to date for debian:10
 ---> dc2eddc15825
Step 2/3 : RUN apt-get update &&     apt-get install sudo
 ---> Running in e65f950f932d
Get:1 http://deb.debian.org/debian buster InRelease [122 kB]
Get:2 http://deb.debian.org/debian buster-updates InRelease [51.9 kB]
Get:3 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:4 http://deb.debian.org/debian buster/main amd64 Packages [7907 kB]
Get:5 http://deb.debian.org/debian buster-updates/main amd64 Packages [9504 B]
Get:6 http://security.debian.org/debian-security buster/updates/main amd64 Packages [269 kB]
Fetched 8424 kB in 1s (5845 kB/s)
Reading package lists...
Building dependency tree...
Reading state information...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  lsb-base
The following NEW packages will be installed:
  lsb-base sudo
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 1273 kB of archives.
After this operation, 3932 kB of additional disk space will be used.
Do you want to continue? [Y/n] Abort.
Do you want to continue? [Y/n] Abort.

build failed: cannot build base image: The command '/bin/sh -c apt-get update &&     apt-get install sudo' returned a non-zero code: 1Do you want to continue? [Y/n] Abort.

Error: build failed: cannot build base image: The command '/bin/sh -c apt-get update &&     apt-get install sudo' returned a non-zero code: 1
```

**Comment by kevinhq** [comment]
but, using default docker image from gitpod (which is ubuntu?), sudo inside runtime works. 

well, i want to have a workspace with debian 10 and sudo. so,this is still restricted? From the log, it shows Gitpod aborts the sudo installation.

**Comment by jankeromnes** [comment]
@kevinhq and @ericsupreme could you please open new issues with your specific use cases? Both sudo in debian and installing jq should definitely work (or be made to work).

**Comment by jankeromnes** [comment]
@kevinhq 

> Do you want to continue? [Y/n] Abort.

This sounds like your build is waiting for interactivity (not supported in Docker builds) and then times out.

Try adding `-y` in `apt-get install -y sudo`.

@ericsupreme could you please share a link to your repository where you try to install `jq`?

**Comment by kevinhq** [comment]
> yeah i'm in the same boat as kevinhq i can't get sudo to work @ all with my current docker image FROM ubuntu:20.04 and it;s a pre built docker image and i just need sudo rights to add jq cli tool. what a royal pain...

I manage to get it work. It's my own mistake. see my updated comment above.

@jankeromnes - exactly as you said. sorry.

**Comment by kevinhq** [comment]
More issue (I don't know if this is how it works or not): After installing some packages with sudo, then close the workspace (workspace is pinned, btw), then here's what happen when I open the same workspace again:

All packages I installed using sudo on previous day are now gone.

Is this an expected result? It would be great if all packages installed using sudo during runtime are retained - even after the workspace is closed and stopped.

Otherwise, having sudo doesn't make much difference. 

We still need to put everything into Dockerfile to make the packages installed and ready - when we start the stopped workpace in another day.

**Comment by jrdzha** [comment]
How can we get sudo work for self-hosted versions of gitpod? We are currently running 0.6.0, and it seems like it's not working here.

**Comment by kcelestinomaria** [comment]
Hello, I am trying to root into a postgresql database in a gitpod workspace so that I can test my web application before refactoring and deploying. I have tried rooting but gitpod asks me for a gitpod password which I don't have. I have used my github's password which I created an account with gitpod but it does not work. Could this feature be enabled , or how could I run a postgresql server on gitpod without making it complicated, thank you.
